httpservice/http_service.go

```diff
@@ -0,0 +1,148 @@
+package httpservice
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/remind101/pkg/retry"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// HTTPService provides a way to make http calls with an aggressive transport
+type HTTPService interface {
+	Call(method string, path string, jsonData interface{}) ([]byte, error)
+	CallWithRetries(method string, path string, jsonData interface{}) ([]byte, error)
+}
+
+// HTTPClient contains the retrier, baseURL and the actual httpClient.
+type HTTPClient struct {
+	client      *http.Client
+	baseURL     *url.URL
+	retrier     *retry.Retrier
+	serviceName string
+}
+
+// NewHTTPClient returns a new HTTPClient for the given service name and the base url
+// The baseURL will be used to prefix the paths provided in Call and CallWithRetries
+func NewHTTPClient(serviceName string, baseURL *url.URL) *HTTPClient {
+	aggressiveTransport := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		Dial: (&net.Dialer{
+			Timeout:   1 * time.Second, // !!!
+			KeepAlive: 90 * time.Second,
+		}).Dial,
+		TLSHandshakeTimeout: 3 * time.Second,
+	}
+
+	return &HTTPClient{
+		baseURL: baseURL,
+		client: &http.Client{
+			Transport: aggressiveTransport,
+		},
+		retrier: retry.NewErrorTypeRetrier(serviceName,
+			retry.DefaultBackOffOpts,
+			(*net.OpError)(nil),
+			(*RetryableHTTPError)(nil)),
+	}
+}
+
+// CallWithRetries makes calls to the network with the default retry logic
+func (client *HTTPClient) CallWithRetries(method string, path string, jsonData interface{}) ([]byte, error) {
+	res, err := client.retrier.Retry(func() (interface{}, error) {
+		return client.Call(method, path, jsonData)
+	})
+	if err != nil {
+		return []byte{}, err
+	}
+	return res.([]byte), nil
+}
+
+// Call formats the network request and returns the raw data as the response
+func (client *HTTPClient) Call(method string, path string, jsonData interface{}) ([]byte, error) {
```

Comment:
Why not just make this more generate and have it implement the `Do` method of `http.Client`?